### PR TITLE
FIP-82b: Configure roles so TribalCouncil can operate

### PR DESCRIPTION
**Pull request by thomas-waite**
# Summary
Configures roles and contract admins throughout the system so that the TribalCouncil can operate the protocol on an operational basis. Specifically:
- Transfers admin of all non-major TribalRoles to the `ROLE_ADMIN`. `ROLE_ADMIN` is held by the TribalCouncil, which will allow it to grant and revoke all non-major roles
- According to an initial spec, creates new roles to act as the contract admin of certain contracts. Sets the admin of these new roles to be the ROLE_ADMIN calls the individual contracts to set the `CONTRACT_ADMIN_ROLE`
- Grants newly created operational roles to the TribalCouncil timelock
- If a contract admin was changed, then it grants the new contract admin role to whichever contract used to be granted the previous contract admin role. This prevents any bricking of the system, any previous admin contracts will still have admin functionality

### FuseGuardian
Prev: TRIBAL_CHIEF_ADMIN_ROLE
New admin role: FUSE_ADMIN
Granted new to: `optimisticTimelock`,`tribalChiefSyncV2`, 'tribalCouncilTimelock`

### optimisticMinter
Prev: GOVERN_ROLE
New: FEI_MINT_ADMIN
Granted new to: 'feiDAOTimelock', 'tribalCouncilTimelock`

### pcvEquityMinter
Prev: GOVERN_ROLE
New: FEI_MINT_ADMIN
Granted new to: 'feiDAOTimelock', 'tribalCouncilTimelock'

### ethLidoPCVDeposit
Implementation on chain does not use a `CoreRef` implementation that has a `setContractAdminRole()` function. Removed.

### indexDelegator
Prev: METAGOVERNANCE_VOTE_ADMIN
New: METAGOVERNANCE_VOTE_ADMIN (no change)
Granted new to: `feiDAOTimelock', 'opsOptimisticTimelock', 'tribalCouncilTimelock`

### ethTokemakPCVDeposit
Unknown admin role: 0x6c9ecf07a5886fd74a8d32f4d3c317a7d5e5b5c7a073a3ab06c217e9ce5288e3
Prev: TOKEMAK_DEPOSIT_ADMIN_ROLE
New: TOKEMAK_DEPOSIT_ADMIN_ROLE (no change)
Granted to: `feiDAOTimelock`, `optimisticTimelock`, `tribalCouncilTimelock`

### uniswapPCVDeposit
Prev: GOVERN_ROLE
New:  PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### daiPSMFeiSkimmer
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### lusdPSMFeiSkimmer
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### ethPSMFeiSkimmer
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### aaveEthPCVDripController
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### daiPCVDripController
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### lusdPCVDripController
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### compoundEthPCVDripController
Prev: GOVERN_ROLE
New: PCV_MINOR_PARAM_ROLE
Granted new to: 'feiDAOTimelock', `optimisticTimelock`,`tribalCouncilTimelock`

### CODE REVIEW

File: contracts/core/TribeRoles.sol [hunk 1]
```diff
@@ -52,12 +52,20 @@ library TribeRoles {
...
+    bytes32 internal constant PSM_ADMIN_ROLE =
+        keccak256("PSM_ADMIN_ROLE");
 
     /// @notice admin of all Minor Roles
     bytes32 internal constant MINOR_ROLE_ADMIN = keccak256("MINOR_ROLE_ADMIN");
 
     /// @notice admin of the Fuse protocol
     bytes32 internal constant FUSE_ADMIN = keccak256("FUSE_ADMIN");
+    /// @notice admin of minting Fei for specific scoped contracts
+    bytes32 internal constant FEI_MINT_ADMIN = keccak256("FUSE_MINT_ADMIN");
```
Comment (thomas-waite) [inline contracts/core/TribeRoles.sol:65]: Push typo fix

File: proposals/dao/fip_82b.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  await validateTransferredRoleAdmins(contracts.core);
+  await validateNewCouncilRoles(contracts.core);
+  await validateContractAdmins(contracts);
+  await validateTribalCouncilRoles(contracts.core, addresses.tribalCouncilTimelock);
+};
+
+/// Validate that all non-major TribeRoles have had their admins transferred to the ROLE_ADMIN
+const validateTransferredRoleAdmins = async (core: Contract) => {
+  const ROLE_ADMIN = ethers.utils.id('ROLE_ADMIN');
+
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:61]: Add in roles that don't yet have an admin

File: proposals/dao/fip_82b.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  expect(await core.getRoleAdmin(ethers.utils.id('ORACLE_ADMIN_ROLE'))).to.be.equal(ROLE_ADMIN);
+  expect(await core.getRoleAdmin(ethers.utils.id('TRIBAL_CHIEF_ADMIN_ROLE'))).to.be.equal(ROLE_ADMIN);
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:63]: Document every role and what it can do

File: proposals/dao/fip_82b.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+/// Validate that the expected new TribeRoles have been created
+const validateNewCouncilRoles = async (core: Contract) => {
+  const ROLE_ADMIN = ethers.utils.id('ROLE_ADMIN');
+
+  expect(await core.getRoleAdmin(ethers.utils.id('FUSE_ADMIN'))).to.be.equal(ROLE_ADMIN);
+  expect(await core.getRoleAdmin(ethers.utils.id('FEI_MINT_ADMIN'))).to.be.equal(ROLE_ADMIN);
+  expect(await core.getRoleAdmin(ethers.utils.id('PCV_ADMIN'))).to.be.equal(ROLE_ADMIN);
+  expect(await core.getRoleAdmin(ethers.utils.id('PSD_ADMIN'))).to.be.equal(ROLE_ADMIN);
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:79]: Typo

File: proposals/dao/fip_82b.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  expect(await core.getRoleAdmin(ethers.utils.id('VOTIUM_ADMIN_ROLE'))).to.be.equal(ROLE_ADMIN);
+};
+
+/// Validate that the expected new TribeRoles have been created
+const validateNewCouncilRoles = async (core: Contract) => {
+  const ROLE_ADMIN = ethers.utils.id('ROLE_ADMIN');
+
+  expect(await core.getRoleAdmin(ethers.utils.id('FUSE_ADMIN'))).to.be.equal(ROLE_ADMIN);
+  expect(await core.getRoleAdmin(ethers.utils.id('FEI_MINT_ADMIN'))).to.be.equal(ROLE_ADMIN);
+  expect(await core.getRoleAdmin(ethers.utils.id('PCV_ADMIN'))).to.be.equal(ROLE_ADMIN);
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:78]: Rename `PCV_ADMIN` to `PCV_MINOR_PARAM_ROLE`

File: proposals/dao/fip_82b.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  expect(await core.getRoleAdmin(ethers.utils.id('PSD_ADMIN'))).to.be.equal(ROLE_ADMIN);
+};
+
+/// Validate that the relevant contract admins have been set to their expected values
+const validateContractAdmins = async (contracts: NamedContracts) => {
+  expect(await contracts.fuseGuardian.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('FUSE_ADMIN'));
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:84]: if (change admin of contract) {
  every contract that called into an contract who's admin has changed, needs the new admin
}

Comment (thomas-waite) [reply proposals/dao/fip_82b.ts:84]: Checkout `permissions.ts`

File: proposals/dao/fip_82b.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+  expect(await contracts.optimisticMinter.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('FEI_MINT_ADMIN'));
+  expect(await contracts.pcvEquityMinter.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('FEI_MINT_ADMIN'));
+
+  expect(await contracts.ethLidoPCVDeposit.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+  expect(await contracts.indexDelegator.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+  expect(await contracts.ethTokemakPCVDeposit.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+  expect(await contracts.uniswapPCVDeposit.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+
+  expect(await contracts.daiPSMFeiSkimmer.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:94]: Note: PCV_ADMIN needs revisiting

File: proposals/dao/fip_82b.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  expect(await contracts.daiPCVDripController.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+  expect(await contracts.lusdPCVDripController.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+  expect(await contracts.compoundEthPCVDripController.CONTRACT_ADMIN_ROLE()).to.be.equal(ethers.utils.id('PCV_ADMIN'));
+};
+
+const validateTribalCouncilRoles = async (core: Contract, tribalCouncilTimelockAddress: string) => {
+  expect(await core.hasRole(ethers.utils.id('FUSE_ADMIN'), tribalCouncilTimelockAddress)).to.be.true;
+  expect(await core.hasRole(ethers.utils.id('FEI_MINT_ADMIN'), tribalCouncilTimelockAddress)).to.be.true;
+  expect(await core.hasRole(ethers.utils.id('PCV_ADMIN'), tribalCouncilTimelockAddress)).to.be.true;
+  expect(await core.hasRole(ethers.utils.id('PCV_GUARDIAN_ADMIN_ROLE'), tribalCouncilTimelockAddress)).to.be.true;
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:108]: Check `PCV_GUARDIAN_ADMIN_ROLE` abilities. Is it worth the risk for the convenience?

File: proposals/dao/fip_82b.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  expect(await core.hasRole(ethers.utils.id('ORACLE_ADMIN_ROLE'), tribalCouncilTimelockAddress)).to.be.true;
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:109]: `ORACLE_ADMIN_ROLE` is another high risk one. Maybe worth keeping

File: proposals/dao/fip_82b.ts (hunk 8 above)
Comment (thomas-waite) [reply proposals/dao/fip_82b.ts:108]: Yep likely - the TribalCouncil should be able to whitelist new PCV deposits without assistance from the DAO

File: proposals/dao/fip_82b.ts (hunk 5 above)
Comment (thomas-waite) [reply proposals/dao/fip_82b.ts:78]: done

File: proposals/dao/fip_82b.ts (hunk 8 above)
Comment (ditchfieldcaleb) [reply proposals/dao/fip_82b.ts:125]: ^ this is a pretty huge power increase and should be discussed more imo

File: proposals/dao/fip_82b.ts (hunk 2 above)
Comment (thomas-waite) [reply proposals/dao/fip_82b.ts:61]: done

File: contracts/core/TribeRoles.sol [hunk 10]
```diff
@@ -48,19 +48,28 @@ library TribeRoles {
...
+    bytes32 internal constant PSM_ADMIN_ROLE = keccak256("PSM_ADMIN_ROLE");
+
     /// @notice admin of all Minor Roles
     bytes32 internal constant MINOR_ROLE_ADMIN = keccak256("MINOR_ROLE_ADMIN");
 
     /// @notice admin of the Fuse protocol
     bytes32 internal constant FUSE_ADMIN = keccak256("FUSE_ADMIN");
 
-    /// @notice capable of vetoing DAO votes or optimistic timelocks
-    bytes32 internal constant VETO_ADMIN = keccak256("VETO_ADMIN");
```
Comment (thomas-waite) [inline contracts/core/TribeRoles.sol:63]: This was a pre-existing role from before the governance upgrade. We now have `POD_VETO_ADMIN` instead

File: scripts/utils/constructProposal.ts [hunk 11]
```diff
@@ -19,7 +19,7 @@ export default async function constructProposal(
   const proposalDescription = proposalInfo.description;
 
   const proposalBuilder = proposals.builders.alpha();
-  proposalBuilder.maxActions = 40;
```
Comment (thomas-waite) [inline scripts/utils/constructProposal.ts:22]: Necessary, don't believe setting to 40 was for a hard reason. Heuristic

File: test/integration/proposals_config.ts [hunk 12]
```diff
@@ -1,12 +1,22 @@
...
+import fip_82b from '@proposals/description/fip_82b';
 
 const proposals: ProposalsConfigMap = {
   fip_82: {
     deploy: true, // deploy flag for whether to run deploy action during e2e tests or use mainnet state
     totalValue: 0, // amount of ETH to send to DAO execution
-    proposal: fip_82, // full proposal file, imported from '@proposals/description/fip_xx.ts'
+    proposal: fip_82,
+    proposalId: '',
+    affectedContractSignoff: [''],
```
Comment (thomas-waite) [inline test/integration/proposals_config.ts:12]: TODO: Complete affectedContractSignoff

File: proposals/dao/fip_82b.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  console.log(`No actions to complete in validate for fip${fipNumber}`);
+  await validateTransferredRoleAdmins(contracts.core);
+  await validateNewCouncilRoles(contracts.core);
+  await validateContractAdmins(contracts);
+  await validateTribalCouncilRoles(contracts.core, addresses.tribalCouncilTimelock);
+  await validateCallingContractsHaveNewAdmin(contracts.core, addresses);
+};
+
+/// Validate that all non-major TribeRoles have had their admins transferred to the ROLE_ADMIN
+const validateTransferredRoleAdmins = async (core: Contract) => {
```
Comment (thomas-waite) [inline proposals/dao/fip_82b.ts:60]: TODO: Create all other TribeRoles that aren't yet existing?

File: protocol-configuration/permissions.ts [hunk 14]
```diff
@@ -12,7 +12,7 @@ export const permissions = {
     'balancerDepositFeiWeth'
   ],
   BURNER_ROLE: [],
-  GOVERN_ROLE: ['core', 'feiDAOTimelock', 'rariTimelock', 'roleBastion'],
```
Comment (thomas-waite) [inline protocol-configuration/permissions.ts:15]: rariTimelock will be removed when I merge in develop

**Review by ditchfieldcaleb — APPROVED** [review]

File: proposals/description/fip_82b.ts [hunk 15]
```diff
@@ -0,0 +1,480 @@
...
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0xc7be67d109cc181c862ac2eaed91ab54817910d73264a8ad5d87592516929a15', // MINOR_ROLE_ADMIN
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Transfer MINOR_ROLE_ADMIN role admin to ROLE_ADMIN'
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:25]: I believe MINOR_ROLE_ADMIN was an old role idea that was meant to be equal to ROLE_ADMIN. If you don't see a need for this role, we can remove it

File: proposals/description/fip_82b.ts [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0x498a9dae57f391d8efcc7bb3e7440ad6a25b1261044ef1b555c5484cb9f67659', // MINTER_ADMIN
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Transfer MINTER_ADMIN role admin to ROLE_ADMIN'
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:35]: weren't we going to rename MINTER_ADMIN so it wasn't confused with MINTER_ROLE. perhaps RATE_LIMITED_MINTER_ADMIN or something

File: protocol-configuration/permissions.ts [hunk 17]
```diff
@@ -24,18 +24,27 @@ export const permissions = {
...
   TRIBAL_CHIEF_ADMIN_ROLE: ['optimisticTimelock', 'tribalChiefSyncV2'],
+  FUSE_ADMIN: ['optimisticTimelock', 'tribalChiefSyncV2'],
   VOTIUM_ADMIN_ROLE: ['opsOptimisticTimelock', 'protocolPodTimelock'],
   PCV_GUARDIAN_ADMIN_ROLE: ['optimisticTimelock'],
   METAGOVERNANCE_VOTE_ADMIN: ['feiDAOTimelock', 'opsOptimisticTimelock'],
   METAGOVERNANCE_TOKEN_STAKING: ['feiDAOTimelock', 'opsOptimisticTimelock'],
   METAGOVERNANCE_GAUGE_ADMIN: ['feiDAOTimelock', 'optimisticTimelock'],
   ROLE_ADMIN: ['feiDAOTimelock', 'tribalCouncilTimelock'],
   POD_METADATA_REGISTER_ROLE: ['tribalCouncilSafe'],
+  FEI_MINT_ADMIN: ['core', 'feiDAOTimelock', 'roleBastion'],
```
Comment (Joeysantoro) [inline protocol-configuration/permissions.ts:45]: why is core a FEI_MINT_ADMIN

File: protocol-configuration/permissions.ts [hunk 18]
```diff
@@ -24,18 +24,27 @@ export const permissions = {
...
   POD_VETO_ADMIN: ['nopeDAO'],
-  POD_ADMIN: ['tribalCouncilTimelock', 'podFactory']
+  POD_ADMIN: ['tribalCouncilTimelock', 'podFactory'],
+  PCV_MINOR_PARAM_ROLE: ['core', 'feiDAOTimelock', 'roleBastion', 'optimisticTimelock', 'opsOptimisticTimelock'],
```
Comment (Joeysantoro) [inline protocol-configuration/permissions.ts:48]: why is core a PCV_MINOR_PARAM_ROLE

Comment (Joeysantoro) [reply protocol-configuration/permissions.ts:48]: Ops timelock maybe shouldn't have this role either? what can it do

File: proposals/description/fip_82b.ts (hunk 15 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:25]: Yep, hasn't been created yet either. Removing

File: protocol-configuration/permissions.ts (hunk 17 above)
Comment (thomas-waite) [reply protocol-configuration/permissions.ts:46]: Shouldn't be, have removed from this permissions mapping. Shouldn't be in there - will make sure the e2e tests don't expect core to be a FEI_MINT_ADMIN

File: protocol-configuration/permissions.ts (hunk 18 above)
Comment (thomas-waite) [reply protocol-configuration/permissions.ts:49]: This `permissions.ts` wasn't up to date for this review, have updated

File: proposals/description/fip_82b.ts (hunk 16 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:25]: Nice, done

File: proposals/dao/fip_82b.ts (hunk 8 above)
Comment (thomas-waite) [reply proposals/dao/fip_82b.ts:138]: This role is having the movement of PCV aspect removed `PCV_GUARDIAN_ADMIN_ROLE` and broken out into a `PCV_MOVER` role. TribalCouncil is getting granted both

File: contracts/core/TribeRoles.sol [hunk 19]
```diff
@@ -48,20 +48,23 @@ library TribeRoles {
...
     bytes32 internal constant TRIBAL_CHIEF_ADMIN = keccak256("TRIBAL_CHIEF_ADMIN_ROLE");
 
+    /// @notice admin of the Tokemak PCV deposits
+    bytes32 internal constant TOKEMAK_DEPOSIT_ADMIN_ROLE = keccak256("TOKEMAK_DEPOSIT_ADMIN_ROLE");
+
     /// @notice admin of PCVGuardian
     bytes32 internal constant PCV_GUARDIAN_ADMIN = keccak256("PCV_GUARDIAN_ADMIN_ROLE");
 
-    /// @notice admin of all Minor Roles
-    bytes32 internal constant MINOR_ROLE_ADMIN = keccak256("MINOR_ROLE_ADMIN");
```
Comment (thomas-waite) [inline contracts/core/TribeRoles.sol:55]: Removed, as this is what `ROLE_ADMIN` does

**Comment by Joeysantoro** [comment]
Made this jist clarifying the PR description accuracy https://gist.github.com/Joeysantoro/21d92cbea44bde4476814952424427fb/revisions

**Comment by Joeysantoro** [comment]
Before we can submit I'd like to see these new role changes reflected in the governance and roles doc https://fei-protocol.github.io/docs/docs/protocol/Governance/detailedAccessControl#fuse_admin

File: proposals/description/fip_82b.ts [hunk 20]
```diff
@@ -0,0 +1,463 @@
...
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0xc307c44629779eb8fc0b85f224c3d22f5876a6c84de0ee42d481eb7814f0d3a8', // ORACLE_ADMIN
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Transfer ORACLE_ADMIN role admin from GOVERNOR to ROLE_ADMIN'
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:45]: nit: technically the role is ORACLE_ADMIN_ROLE. i.e. the comments are off but args correct

File: proposals/description/fip_82b.ts [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+const fip_82b: ProposalDescription = {
+  title: 'FIP-82b: Authorise the TribalCouncil with necessary roles',
+  commands: [
+    /////// Transfer all non-major role admins to the ROLE_ADMIN, to allow TribalCouncil to manage ////////
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0xf0b50f04623eeaacfa1f202e062a3001c925a35c6b75d6903e67b43f44bbf152', // PARAMETER_ADMIN
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:12]: PARAMETER_ADMIN doesn't appear anywhere in PCV. or the permissions test. I prefer removing it

File: proposals/description/fip_82b.ts [hunk 22]
```diff
@@ -0,0 +1,463 @@
...
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Transfer PARAMETER_ADMIN role admin to ROLE_ADMIN'
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0xdc81827f5af6c7965785d62c38ca3481ccf540d7f561cac825518e49d6161c95', // RATE_LIMITED_MINTER_ADMIN
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:22]: RATE_LIMITED_MINTER_ADMIN doesn't appear anywhere in protocol or the permissions test.

However we would plan to use it with GlobalFeiMinter, so I'd just want to see it in permissions.ts with an empty list

File: proposals/description/fip_82b.ts [hunk 23]
```diff
@@ -0,0 +1,463 @@
...
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Transfer RATE_LIMITED_MINTER_ADMIN role admin to ROLE_ADMIN'
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0x6cad5c16d973906992d8e1412f4aefeca4ae44220c203bbcd6e8e66e7c717be0', // OPTIMISTIC_ADMIN
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:32]: OPTIMISTIC_ADMIN doesn't appear anywhere in the protocol or the permissions test, and I don't think we have plans to use it. I prefer removing it from TribeRoles and this script.

I believe this was originally meant to work like POD_ADMIN

File: proposals/description/fip_82b.ts [hunk 24]
```diff
@@ -0,0 +1,463 @@
...
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Transfer TRIBAL_CHIEF_ADMIN role admin from GOVERNOR to ROLE_ADMIN'
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0xdf6ce05acd28d71e96472375ef55c4a692f167af3ccda9500570f7373c1ba22c', // PCV_GUARDIAN_ADMIN
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:62]: nit: technically the role is PCV_GUARDIAN_ADMIN_ROLE. i.e. the comments are off but args correct

File: proposals/description/fip_82b.ts [hunk 25]
```diff
@@ -0,0 +1,463 @@
...
+    },
+    {
+      target: 'core',
+      values: '0',
+      method: 'createRole(bytes32,bytes32)',
+      arguments: [
+        '0x64bdfa4de883e5fe4519e80a554fbf0d63c0f6271ac8b89b18edcc9ef151ad60', // MINOR_PARAM_ROLE
+        '0x2172861495e7b85edac73e3cd5fbb42dd675baadf627720e687bcfdaca025096' // ROLE_ADMIN
+      ],
+      description: 'Create MINOR_PARAM_ROLE role, assigning ROLE_ADMIN as the role admin'
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:166]: MINOR_PARAM_ROLE doesn't appear anywhere in the protocol or the permissions test.

Either this or PCV_MINOR_PARAM_ROLE should exist but not both. The other should be removed.

File: proposals/description/fip_82b.ts [hunk 26]
```diff
@@ -0,0 +1,463 @@
...
+      ],
+      description: 'Revoke GUARDIAN from old pcvGuardian'
+    }
+  ],
+  description: `
+    FIP_82b: Create, grant and make the TribalCouncil admin of various access control roles to allow
+    it to manage the protocol.
+
+    These roles allow the TribalCouncil to authorise pods with the necessary access control to operate 
+    distinct parts of the protocol.
```
Comment (Joeysantoro) [inline proposals/description/fip_82b.ts:459]: Please link the PR, snapshot and tribe forum post in the description

File: proposals/description/fip_82b.ts (hunk 24 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:62]: updated

File: proposals/description/fip_82b.ts (hunk 23 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:32]: Makes sense, removed

File: proposals/description/fip_82b.ts (hunk 22 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:22]: I test that it is correctly created and in the `dao/fip_82b.ts` validate script, will add to the permissions test.

File: proposals/description/fip_82b.ts (hunk 21 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:12]: It's used in the `NonCustodialPSM`, but hasn't been created or granted out yet. We should leave as it's in an access modifier on `NonCustodialPSM`

File: proposals/description/fip_82b.ts (hunk 20 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:45]: done

File: proposals/description/fip_82b.ts (hunk 25 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:166]: Removing MINOR_PARAM_ROLE, prefer having a topic scoped role like `PCV_MINOR_PARAM_ROLE` around

File: proposals/description/fip_82b.ts (hunk 26 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:459]: Done, and fleshed out the description

**Comment by thomas-waite** [comment]
> Made this jist clarifying the PR description accuracy https://gist.github.com/Joeysantoro/21d92cbea44bde4476814952424427fb/revisions

Updated PR description with this and made the two role changes in there

**Comment by thomas-waite** [comment]
> Before we can submit I'd like to see these new role changes reflected in the governance and roles doc https://fei-protocol.github.io/docs/docs/protocol/Governance/detailedAccessControl#fuse_admin

Updated

File: proposals/description/fip_82b.ts (hunk 21 above)
Comment (Joeysantoro) [reply proposals/description/fip_82b.ts:12]: imo the roles should be consolidated/clarified. There is a PSM_ADMIN_ROLE and a MINOR_PARAMETER_ADMIN. At least one of those should replace this one or vice versa

**Review by Joeysantoro — APPROVED** [review]

File: proposals/description/fip_82b.ts (hunk 21 above)
Comment (thomas-waite) [reply proposals/description/fip_82b.ts:12]: As discussed, leaving for now

**Review by Joeysantoro — APPROVED** [review]

**Pull request merged**